aggregator/src/aggregator.rs

```diff
@@ -2024,29 +2025,36 @@ impl VdafOps {
         // that the task IDs match as this should be guaranteed by the caller.
         let req = Arc::new(CollectReq::<Q>::get_decoded(req_bytes)?);
         assert_eq!(req.task_id(), task.id());
-
-        let aggregation_param =
-            A::AggregationParam::get_decoded(req.as_ref().aggregation_parameter())?;
-
-        // Check that the batch interval is valid for the task
-        // https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-4.5.6.1.1
-        if !Q::validate_collect_identifier(&task, req.query().batch_identifier()) {
-            return Err(Error::BatchInvalid(
-                *task.id(),
-                format!("{}", req.query().batch_identifier()),
-            ));
-        }
+        let aggregation_param = Arc::new(A::AggregationParam::get_decoded(
+            req.as_ref().aggregation_parameter(),
+        )?);
 
         Ok(datastore
             .run_tx(move |tx| {
-                let aggregation_param = aggregation_param.clone();
-                let task = task.clone();
-                let req = req.clone();
+                let (task, req, aggregation_param) = (
+                    Arc::clone(&task),
+                    Arc::clone(&req),
+                    Arc::clone(&aggregation_param),
+                );
                 Box::pin(async move {
+                    let batch_identifier = Q::batch_identifier_for_query(tx, &task, req.query())
+                        .await?
+                        .ok_or_else(|| {
+                            datastore::Error::User(anyhow!("No batch ready for collection").into())
```

Comment:
I'd prefer not to use `anyhow` here. Something like `aggregator::Error::BatchInvalid(*task.id(), "no batch ready for collection".to_string())` should work.